### PR TITLE
bazel: use transitions to apply cross-compile platform automatically to oci_image

### PR DESCRIPTION
**Pull request by Strum355**
Removes the need to pass `--config=docker-darwin` through the following mechanisms:

1. `--enable_platform_specific_config` to enable certain flags on macos only e.g. `--extra_toolchains @zig_sdk//toolchain:linux_amd64_gnu.2.34` and `--sandbox_add_mount_pair=/tmp` (see [.bazelrc change](https://github.com/sourcegraph/sourcegraph/pull/60569/files?file-filters%5B%5D=dotfile&show-viewed-files=true))
2. Apply a transition (using https://github.com/fmeum/with_cfg.bzl, please view [the following great video on it](https://www.youtube.com/watch?v=U5bdQRQY-io)) on `oci_image` targets when on the `@platforms//os:macos` platform to transition to the `@zig_sdk//platform:linux_amd64` platform. 
	- This will start at `oci_image` targets and propagate down to e.g. `go_{binary,library}` etc targets with the "transitioned" platform configuration, resulting in them being built with the transitioned-to platform
3. Remove `darwin_docker_e2e_go` config_setting and `darwin-docker` bool_flag.
	- These aren't necessary anymore, as the places where these were used were not in the transitive closure rooted at an `oci_image` target, meaning they wouldn't be transitioned.

To review, view [the following (filtered) files](https://github.com/sourcegraph/sourcegraph/pull/60569/files?file-filters%5B%5D=.bzl&file-filters%5B%5D=.sh&file-filters%5B%5D=.yaml&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile&show-viewed-files=true)  along with [the root BUILD.bazel](https://github.com/sourcegraph/sourcegraph/pull/60569/files#diff-7fc57714ef13c3325ce2a1130202edced92fcccc0c6db34a72f7b57f60d552a3). All the other files are just changing the `load` statements from `@rules_oci` to `//dev:oci_defs.bzl`

## Test plan

CI, checked image locally and `sg test bazel-backend-integration` & `sg test bazel-e2e`

### CODE REVIEW

File: sg.config.yaml [hunk 1]
```diff
@@ -637,13 +637,8 @@ commands:
...
     install: |
-      if [[ $(uname) == "Linux" ]]; then
-        bazel build //cmd/executor-kubernetes:image_tarball
-        docker load --input $(bazel cquery //cmd/executor-kubernetes:image_tarball --output=files)
-      else
-        bazel build //cmd/executor-kubernetes:image_tarball --config darwin-docker
-        docker load --input $(bazel cquery //cmd/executor-kubernetes:image_tarball --config darwin-docker --output=files)
-      fi
+      bazel build //cmd/executor-kubernetes:image_tarball
+      docker load --input $(bazel cquery //cmd/executor-kubernetes:image_tarball --output=files)
```
Comment (Strum355) [inline sg.config.yaml:641]: Should be able to replace with `bazel run //cmd/executor-kubernetes:image_tarball` but Ive observed an error with `CopyFile` action. Will investigate after this PR

File: dev/oci_defs.bzl [hunk 2]
```diff
@@ -18,5 +19,11 @@ def oci_tarball(name, **kwargs):
...
     )
 
-oci_image = _oci_image
+_oci_image_builder = with_cfg(_oci_image)
+_oci_image_builder.set("platforms", select({
+    "@platforms//os:macos": [Label("@zig_sdk//platform:linux_amd64")],
+    "//conditions:default": [],
+}))
+
+oci_image, _oci_image_internal = _oci_image_builder.build()
```
Comment (Strum355) [inline dev/oci_defs.bzl:28]: The magic we perform here adds some extra indirection that affects query commands. Ill try document them here before merging

Comment (jhchabran) [reply dev/oci_defs.bzl:28]: Good you mentioned to watch the video, otherwise it would be indeed a bit obscure :P

**Review by jhchabran — APPROVED** [review]
This is really cool and will simplify things a lot 🙏 Well done!

**Pull request merged**